torch/csrc/distributed/rpc/rref_context.cpp

```diff
@@ -37,33 +38,117 @@ const std::shared_ptr<RpcAgent>& RRefContext::agent() const {
   return agent_;
 }
 
-void RRefContext::addFork(at::IValue&& value) {
+RRefForkData RRefContext::forkTo(
+    const std::shared_ptr<RRef>& rref, worker_id_t forkDst) {
+  auto forkRequest = rref->fork();
+  if (rref->owner() != forkDst) {
+    // if fork destination if not owner, the forked UserRRef needs to be tracked
+    // properly
+    if (rref->isOwner()) {
+      // fork from owner
+      acceptUserRRef(forkRequest.rrefId_, forkRequest.forkId_, forkDst);
+    } else {
+      // fork from user, rref cannot be destructed until the fork request is
+      // accepted by the owner
+      {
+        std::lock_guard<std::mutex> lock(mutex_);
+        pendingForkRequests_[forkRequest.forkId_] = rref;
+      }
+      // notify owner
+      agent_->send(
+          agent_->getWorkerId(rref->owner()),
+          ScriptForkNotify(forkRequest.toIValue(), forkDst).toMessage()
+      );
+    }
+  }
+  return forkRequest;
+}
+
+void RRefContext::acceptUserRRef(
+    const RRefId& rrefId, const ForkId& forkId, worker_id_t user) {
+  addForkOfOwner(rrefId, forkId);
+  agent_->send(
```

Comment:
Should do something with the result of this send -- for error handling at least.